add step_dummy_multi_label()

EmilHvitfeldt opened the pull request:
This PR is not finished. Working prototype done. Still needs

- Documentation
- examples
- unit tests

EmilHvitfeldt commented:
Should be ready to go:

``` r
library(recipes)

languages <- tribble(
  ~lang_1,    ~lang_2,   ~lang_3,  ~lang_4,
  "English",  "Italian", NA,       NA,
  "Spanish",  NA,        "French", NA,
  "Armenian", "English", "French", NA,
  NA,         NA,        NA,       NA
)

rec <- recipe(~ ., data = languages) %>%
  step_dummy_multi_label(starts_with("lang"), prefix = "lang") %>% 
  prep() 

tidy(rec, 1)
#> # A tibble: 5 x 3
#>   terms  columns  id                     
#>   <chr>  <chr>    <chr>                  
#> 1 lang_1 Armenian dummy_multi_label_6xSCl
#> 2 lang_1 English  dummy_multi_label_6xSCl
#> 3 lang_1 French   dummy_multi_label_6xSCl
#> 4 lang_1 Italian  dummy_multi_label_6xSCl
#> 5 lang_1 Spanish  dummy_multi_label_6xSCl

bake(rec, new_data = NULL)
#> # A tibble: 4 x 5
#>   lang_Armenian lang_English lang_French lang_Italian lang_Spanish
#>           <int>        <int>       <int>        <int>        <int>
#> 1             0            1           0            1            0
#> 2             0            0           1            0            1
#> 3             1            1           1            0            0
#> 4             0            0           0            0            0
```

<sup>Created on 2021-06-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

topepo commented on R/dummy_multi_label.R:
```diff
@@ -0,0 +1,309 @@
+#' multi Label Dummy Variables Creation
+#'
+#' `step_dummy_multi_label` creates a *specification* of a recipe
```
```suggestion
#' `step_dummy_multi_label()` creates a *specification* of a recipe
```

topepo commented on R/dummy_multi_label.R:
```diff
@@ -0,0 +1,309 @@
+#'  role should they be assigned?. By default, the function assumes
+#'  that the binary dummy variable columns created by the original
+#'  variables will be used as predictors in a model.
+#' @param threshold A numeric value between 0 and 1 or an integer greater or
```
This really bit us for something similar that we did with xgboost. I'd say let's just use the fractional value.

topepo commented:
Now that I look at it, I'm wondering if the use of "label" in the name might make people think that it is about the outcome (I realize that I suggested it 😬 ). How about `step_dummy_multi_choice()`?

EmilHvitfeldt commented:
I like `step_dummy_multi_choice()` better. All suggestions have been added along with a NEWS bullet

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,307 @@
+
+  as.matrix(res) %>%
+    as_tibble() %>%
+    mutate_all(as.integer)
```
Do we have any concerns about adding a new use of a superseded function to the package at this point? We could coerce to integer when still in matrix form like `as.matrix(res) %>% apply(2, as.integer) %>% as_tibble()`?

EmilHvitfeldt replied:
done!

EmilHvitfeldt commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,307 @@
+#'  role should they be assigned?. By default, the function assumes
+#'  that the binary dummy variable columns created by the original
+#'  variables will be used as predictors in a model.
+#' @param threshold A numeric value between 0 and 1. If it's less than one then
+#'  factor levels whose rate of occurrence in the training set are below
+#'  `threshold` will be "othered". If it is 0 then no thresholding will be done.
```
I'll add a "See Details" and a paragraph with further explanations.

EmilHvitfeldt commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,307 @@
+#' @param input A character vector containing the names of the columns used.
+#'  This is `NULL` until the step is trained by [prep.recipe()].
+#' @param other A single character value for the "other" category.
+#' @param naming A function that defines the naming convention for
+#'  new dummy columns. See Details below.
```
Would be a good idea to `@inheritParams` or `@template` but I feel it should be done in a separate PR since it would hit multiple other steps. `naming` is used in `step_dummy()` and `step_ratio()`

EmilHvitfeldt replied:
I ended up using `@inheritParams step_dummy`

EmilHvitfeldt commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,307 @@
+#'  not valid variable names (e.g. "some text with spaces"), it will
+#'  be changed by [base::make.names()] to be valid (see the example
+#'  below). The naming format can be changed using the `naming`
+#'  argument and the function [dummy_names()] is the default. This
```
I tried to clarify this a bit. Note that this block was taken from `step_dummy()`
https://github.com/tidymodels/recipes/blob/2f73e8c400351f906494ede06c0b0961b358475c/R/dummy.R#L51-L62

EmilHvitfeldt commented on R/dummy_multi_choice.R:
````diff
@@ -0,0 +1,307 @@
+#'  0       0      0       0       0
+#'  ```
+#'
+#'  The function allows for non-standard naming of the resulting
````
Can you clarify which block you are referring to and what is not appropriate?

topepo replied:
I think I meant the bit about "The function allows for non-standard naming of the resulting"

juliasilge replied:
This bit of documentation is now in `man-roxygen/dummy-naming.R` and says:

> This recipe step allows for flexible naming of the resulting variables. For an unordered factor ...

EmilHvitfeldt commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,307 @@
+  tibble::tibble(
+    name = c("num_comp", "threshold"),
+    call_info = list(
+      list(pkg = "dials", fun = "num_comp", range = c(1L, 4L)),
```
Only removing `num_comp` right?

EmilHvitfeldt commented on R/dummy_multi_choice.R:
````diff
@@ -0,0 +1,307 @@
+#'  NA         NA         NA
+#'  ```
+#'
+#'  would end up looking like this
````
Done!
<img width="658" alt="Screen Shot 2021-08-16 at 9 59 42 PM" src="https://user-images.githubusercontent.com/14034784/129666383-dddb7520-b56d-4db4-b70e-d377f066bad5.png">

EmilHvitfeldt commented:
Alright, I went through and incorporated changes, added pkgdown, and made sure that this step is up to date with respect to

- `@template` usage
- redo of documentation for `tidy.*` steps
- switch to `recipes_eval_select()`

I can open another issue to deal the following comment, as it should be updated across multiple steps.

```suggestion
#'  argument; the function [dummy_names()] is the default. This
```
It's pretty confusing that both `make.names()` and `dummy_names()` are in play here. Can we clarify this?

_Originally posted by @juliasilge in https://github.com/tidymodels/recipes/pull/722#discussion_r648758578_

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#'
+#' @inheritParams step_center
+#' @inheritParams step_dummy
+#' @param ... One or more selector functions to choose which variables will be
+#'  used to compute the components. See [selections()] for more details.
+#' @param role For model terms created by this step, what analysis
+#'  role should they be assigned?. By default, the function assumes
+#'  that the binary dummy variable columns created by the original
+#'  variables will be used as predictors in a model.
```
These should be inherited, I believe.

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#'  role should they be assigned?. By default, the function assumes
+#'  that the binary dummy variable columns created by the original
+#'  variables will be used as predictors in a model.
+#' @param threshold A numeric value between 0 and 1. If less than one,
+#'  factor levels whose rate of occurrence in the training set are below
+#'  `threshold` will be "othered". See Details below.
+#'   If `threshold = 0` then no thresholding is done.
```
Can this be the same as what is in `step_other()` to be consistent, and then inherited? At the least, we should use the same wording.

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#'  factor levels whose rate of occurrence in the training set are below
+#'  `threshold` will be "othered". See Details below.
+#'   If `threshold = 0` then no thresholding is done.
+#' @param levels A list with the information needed to
+#'  create dummy variables for each variable contained in
+#'  `terms`. This is `NULL` until the step is trained by
+#'  [prep.recipe()].
```
This should be inherited.

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#'  [prep.recipe()].
+#' @param input A character vector containing the names of the columns used.
+#'  This is `NULL` until the step is trained by [prep.recipe()].
+#' @param other A single character value for the "other" category.
```
Should inherit from `step_other()`

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#' @param input A character vector containing the names of the columns used.
+#'  This is `NULL` until the step is trained by [prep.recipe()].
+#' @param other A single character value for the "other" category.
+#' @param prefix A character string that will be the prefix to the resulting
+#'  new variables. See notes below.
```
Should inherit from `step_pca()`

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#' @param other A single character value for the "other" category.
+#' @param prefix A character string that will be the prefix to the resulting
+#'  new variables. See notes below.
+#' @param keep_original_cols A logical to keep the original variables in the
+#'  output. Defaults to `FALSE`.
```
Should inherit

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#' @param keep_original_cols A logical to keep the original variables in the
+#'  output. Defaults to `FALSE`.
+#' @template step-return
+#' @keywords datagen
+#' @concept preprocessing
+#' @concept dummy_multi_choice
+#' @concept projection_methods
```
```suggestion
#' @family {dummy variable and encoding steps}
```

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#' @concept projection_methods
+#' @export
+#' @details
+#' `step_dummy_multi_choice()` will create a set of binary dummy
+#'  variables from a selection of factor variables. For data such as:
+#'
+#' | lang_1   | lang_2  | lang_3 |
+#' | -------- | ------- | ------ |
+#' | English  | Italian | NA     |
+#' | Spanish  | NA      | French |
+#' | Armenian | English | French |
+#' | NA       | NA      | NA     |
+#'
+#'  the results will be:
+#'
+#' | Armenian | English | French | Italian | Spanish |
+#' | -------- | ------- | ------ | ------- | ------- |
+#' | 0        | 1       | 0      | 1       | 0       |
+#' | 0        | 0       | 1      | 0       | 1       |
+#' | 1        | 1       | 1      | 0       | 0       |
+#' | 0        | 0       | 0      | 0       | 0       |
```
This is the same as what is in the example; I believe I agree with @topepo when he was saying earlier that this is likely not needed.

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#' | 1        | 1       | 1      | 0       | 0       |
+#' | 0        | 0       | 0      | 0       | 0       |
+#'
+#'  This recipe step allows for flexible naming of the resulting
+#'  variables. For an unordered factor named `x`, with levels `"a"`
+#'  and `"b"`, the default naming convention would be to create a
+#'  new variable called `x_b`. The naming format can be changed using
+#'  the `naming` argument; the function [dummy_names()] is the
+#'  default. This function will also change the names of ordinal
+#'  dummy variables. Instead of values such as "`.L`", "`.Q`", or
+#'  "`^4`", ordinal dummy variables are given simple integer
+#'  suffixes such as "`_1`", "`_2`", etc.
```
Let's create a blurb in `man-roxygen` for this instead of copying/pasting, called maybe `dummy-naming.R`.

EmilHvitfeldt replied:
done, I used the wording from `step_dummy_multi_choice` since it got updated during this PR

juliasilge commented on R/dummy_multi_choice.R:
```diff
@@ -0,0 +1,305 @@
+#'
+#' bake(dummy_multi_choice_rec2, new_data = NULL)
+#' tidy(dummy_multi_choice_rec2, number = 1)
+#' @seealso [step_dummy()]
```
```suggestion
```

EmilHvitfeldt commented:
All done!
